File: src/KRATE.Core/Tools/Colors.cs
using System.Globalization;

namespace Krate.Core;

public static class Colors
{
    /// <summary>Accepts "#3af", "#33aaff", "3aafff", "rgb(51, 170, 255)" or "hsl(204, 100%, 60%)".</summary>
    public static (int R, int G, int B) Parse(string input)
    {
        var s = input.Trim().ToLowerInvariant().Replace(" ", "");

        if (s.StartsWith("rgb"))
        {
            var n = Numbers(s);
            return (Clamp(n[0]), Clamp(n[1]), Clamp(n[2]));
        }
        if (s.StartsWith("hsl"))
        {
            var n = Numbers(s);
            return FromHsl(n[0], n[1] / 100, n[2] / 100);
        }

        var hex = s.TrimStart('#');
        if (hex.Length == 3) hex = string.Concat(hex.Select(c => $"{c}{c}")); // #3af → #33aaff
        if (hex.Length != 6 || !int.TryParse(hex, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var v))
            throw new ArgumentException(Strings.Get("Error_BadColor", input));
        return (v >> 16 & 0xFF, v >> 8 & 0xFF, v & 0xFF);
    }

    /// <summary>The same color in every notation — that's what the tool is actually for.</summary>
    public static string Describe(string input) => Describe(Parse(input));

    public static string Describe(int rgb) => Describe((rgb >> 16 & 0xFF, rgb >> 8 & 0xFF, rgb & 0xFF));

    public static string Describe((int R, int G, int B) c)
    {
        var (h, s, l) = ToHsl(c);
        return string.Join('\n',
            $"HEX  #{c.R:X2}{c.G:X2}{c.B:X2}",
            $"RGB  rgb({c.R}, {c.G}, {c.B})",
            string.Create(CultureInfo.InvariantCulture, $"HSL  hsl({h:0}, {s * 100:0}%, {l * 100:0}%)"));
    }

    public static (double H, double S, double L) ToHsl((int R, int G, int B) c)
    {
        double r = c.R / 255.0, g = c.G / 255.0, b = c.B / 255.0;
        double max = Math.Max(r, Math.Max(g, b)), min = Math.Min(r, Math.Min(g, b));
        var l = (max + min) / 2;
        if (max == min) return (0, 0, l); // grey: hue is undefined, report 0

        var d = max - min;
        var s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
        var h = max == r ? (g - b) / d + (g < b ? 6 : 0)
              : max == g ? (b - r) / d + 2
              : (r - g) / d + 4;
        return (h * 60, s, l);
    }

    public static (int R, int G, int B) FromHsl(double h, double s, double l)
    {
        h = (h % 360 + 360) % 360 / 360;
        s = Math.Clamp(s, 0, 1);
        l = Math.Clamp(l, 0, 1);
        var q = l < 0.5 ? l * (1 + s) : l + s - l * s;
        var p = 2 * l - q;
        return (Channel(p, q, h + 1.0 / 3), Channel(p, q, h), Channel(p, q, h - 1.0 / 3));
    }

    static int Channel(double p, double q, double t)
    {
        t = (t % 1 + 1) % 1;
        var v = t < 1.0 / 6 ? p + (q - p) * 6 * t
              : t < 1.0 / 2 ? q
              : t < 2.0 / 3 ? p + (q - p) * (2.0 / 3 - t) * 6
              : p;
        return (int)Math.Round(v * 255);
    }

    /// <summary>Reads the numbers between the parentheses of an rgb()/hsl() notation.
    ///
    /// Both bounds are validated first: "rgb(0" has no closing paren, so LastIndexOf returned -1
    /// and the range went negative, surfacing .NET's raw "length ('-5') must be a non-negative
    /// value. (Parameter 'length')" to the user. A malformed colour is a bad colour, not a
    /// crash.</summary>
    static double[] Numbers(string s)
    {
        var open = s.IndexOf('(');
        var close = s.LastIndexOf(')');
        if (open < 0 || close <= open) throw new ArgumentException(Strings.Get("Error_BadColor", s));
        return s[(open + 1)..close]
            .Split([',', '/'], StringSplitOptions.RemoveEmptyEntries)
            .Select(p => double.TryParse(p.Trim().TrimEnd('%'), NumberStyles.Float, CultureInfo.InvariantCulture, out var v)
                ? v
                : throw new ArgumentException(Strings.Get("Error_BadColor", s)))
            .ToArray();
    }

    static int Clamp(double v) => (int)Math.Clamp(Math.Round(v), 0, 255);

    static string Hex((int R, int G, int B) c) => $"#{c.R:X2}{c.G:X2}{c.B:X2}";

    /// <summary>Colour harmonies built by rotating the hue. Input is any colour notation.</summary>
    public static string Palette(string input)
    {
        var (h, s, l) = ToHsl(Parse(input));
        string At(double degrees) => Hex(FromHsl(h + degrees, s, l));
        return string.Join('\n',
            $"{Strings.Get("Color_Base"),-16} {At(0)}",
            $"{Strings.Get("Color_Complementary"),-16} {At(180)}",
            $"{Strings.Get("Color_Triadic"),-16} {At(120)}  {At(240)}",
            $"{Strings.Get("Color_Analogous"),-16} {At(-30)}  {At(30)}",
            $"{Strings.Get("Color_SplitComp"),-16} {At(150)}  {At(210)}",
            $"{Strings.Get("Color_Tetradic"),-16} {At(90)}  {At(180)}  {At(270)}");
    }

    /// <summary>WCAG contrast ratio between two colours (one per line) and which levels it passes.</summary>
    public static string Contrast(string input)
    {
        var lines = input.Split('\n', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
        if (lines.Length < 2) throw new ArgumentException(Strings.Get("Error_NeedTwoColors"));
        var ratio = ContrastRatio(Parse(lines[0]), Parse(lines[1]));

        string Verdict(double min) => ratio >= min ? Strings.Get("Color_Pass") : Strings.Get("Color_Fail");
        return string.Join('\n',
            string.Create(CultureInfo.InvariantCulture, $"{Strings.Get("Color_Ratio")}  {ratio:0.00}:1"),
            $"AA  ({Strings.Get("Color_NormalText")})   {Verdict(4.5)}",
            $"AA  ({Strings.Get("Color_LargeText")})   {Verdict(3.0)}",
            $"AAA ({Strings.Get("Color_NormalText")})   {Verdict(7.0)}",
            $"AAA ({Strings.Get("Color_LargeText")})   {Verdict(4.5)}");
    }

    /// <summary>WCAG 2.x contrast ratio: (L1 + 0.05) / (L2 + 0.05), lighter over darker.</summary>
    public static double ContrastRatio((int R, int G, int B) a, (int R, int G, int B) b)
    {
        var (la, lb) = (RelativeLuminance(a), RelativeLuminance(b));
        var (hi, lo) = (Math.Max(la, lb), Math.Min(la, lb));
        return (hi + 0.05) / (lo + 0.05);
    }

    static double RelativeLuminance((int R, int G, int B) c)
    {
        // sRGB → linear, per the WCAG definition, then the Rec.709 luminance weights.
        static double Channel(int v)
        {
            var s = v / 255.0;
            return s <= 0.03928 ? s / 12.92 : Math.Pow((s + 0.055) / 1.055, 2.4);
        }
        return 0.2126 * Channel(c.R) + 0.7152 * Channel(c.G) + 0.0722 * Channel(c.B);
    }

    /// <summary>Colour temperature in Kelvin → the approximate RGB of that white point
    /// (candlelight ~1900K, daylight ~6500K). Uses the Tanner Helland approximation.</summary>
    public static string Temperature(string input)
    {
        if (!double.TryParse(input.Trim().TrimEnd('k', 'K'), NumberStyles.Float, CultureInfo.InvariantCulture, out var kelvin))
            throw new ArgumentException(Strings.Get("Error_NeedNumber"));
        if (kelvin is < 1000 or > 40000) throw new ArgumentException(Strings.Get("Error_KelvinRange"));
        var c = KelvinToRgb(kelvin);
        return Describe(c);
    }

    // Standard colour-vision-deficiency simulation matrices (applied in sRGB space).
    static readonly double[] Protanopia = [0.567, 0.433, 0, 0.558, 0.442, 0, 0, 0.242, 0.758];
    static readonly double[] Deuteranopia = [0.625, 0.375, 0, 0.70, 0.30, 0, 0, 0.30, 0.70];
    static readonly double[] Tritanopia = [0.95, 0.05, 0, 0, 0.433, 0.567, 0, 0.475, 0.525];

    /// <summary>Shows how a colour appears under the common colour-blindness types — for checking a
    /// palette stays distinguishable.</summary>
    public static string ColorBlind(string input)
    {
        var c = Parse(input);
        var gray = Clamp(0.299 * c.R + 0.587 * c.G + 0.114 * c.B);
        return string.Join('\n',
            $"{Strings.Get("Cvd_Normal"),-18} {Hex(c)}",
            $"{Strings.Get("Cvd_Protan"),-18} {Hex(Simulate(c, Protanopia))}",
            $"{Strings.Get("Cvd_Deuter"),-18} {Hex(Simulate(c, Deuteranopia))}",
            $"{Strings.Get("Cvd_Tritan"),-18} {Hex(Simulate(c, Tritanopia))}",
            $"{Strings.Get("Cvd_Achroma"),-18} {Hex((gray, gray, gray))}");
    }

    public static (int R, int G, int B) Simulate((int R, int G, int B) c, double[] m) =>
        (Clamp(m[0] * c.R + m[1] * c.G + m[2] * c.B),
         Clamp(m[3] * c.R + m[4] * c.G + m[5] * c.B),
         Clamp(m[6] * c.R + m[7] * c.G + m[8] * c.B));

    public static (int R, int G, int B) SimulateProtanopia((int R, int G, int B) c) => Simulate(c, Protanopia);
    public static (int R, int G, int B) SimulateDeuteranopia((int R, int G, int B) c) => Simulate(c, Deuteranopia);
    public static (int R, int G, int B) SimulateTritanopia((int R, int G, int B) c) => Simulate(c, Tritanopia);

    public static (int R, int G, int B) KelvinToRgb(double kelvin)
    {
        var t = kelvin / 100;
        double r, g, b;

        if (t <= 66) { r = 255; g = 99.4708025861 * Math.Log(t) - 161.1195681661; }
        else { r = 329.698727446 * Math.Pow(t - 60, -0.1332047592); g = 288.1221695283 * Math.Pow(t - 60, -0.0755148492); }

        if (t >= 66) b = 255;
        else if (t <= 19) b = 0;
        else b = 138.5177312231 * Math.Log(t - 10) - 305.0447927307;

        return (Clamp(r), Clamp(g), Clamp(b));
    }
}

File: rust/src/colors.rs
//! Colour notations, harmonies and WCAG contrast. Mirrors `Krate.Core.Colors`.

use crate::i18n;
use crate::tools::round_half_even;

pub type Rgb = (i32, i32, i32);

fn clamp_channel(v: f64) -> i32 {
    round_half_even(v).clamp(0.0, 255.0) as i32
}

/// Reads `#rgb`, `#rrggbb`, `rgb(...)` or `hsl(...)`, in any casing and with spaces anywhere.
pub fn parse(input: &str) -> Result<Rgb, String> {
    let s = input.trim().to_lowercase().replace(' ', "");
    let bad = || i18n::format("Error_BadColor", &[input]);

    if s.starts_with("rgb") || s.starts_with("hsl") {
        let n = inner_numbers(&s).ok_or_else(bad)?;
        if n.len() < 3 {
            return Err(bad());
        }
        return Ok(if s.starts_with("rgb") {
            (clamp_channel(n[0]), clamp_channel(n[1]), clamp_channel(n[2]))
        } else {
            from_hsl(n[0], n[1] / 100.0, n[2] / 100.0)
        });
    }

    let mut hex = s.trim_start_matches('#').to_string();
    if hex.len() == 3 {
        hex = hex.chars().flat_map(|c| [c, c]).collect(); // #3af -> #33aaff
    }
    if hex.len() != 6 {
        return Err(bad());
    }
    let v = u32::from_str_radix(&hex, 16).map_err(|_| bad())?;
    Ok((
        (v >> 16 & 0xFF) as i32,
        (v >> 8 & 0xFF) as i32,
        (v & 0xFF) as i32,
    ))
}

fn inner_numbers(s: &str) -> Option<Vec<f64>> {
    let open = s.find('(')?;
    let close = s.rfind(')')?;
    Some(
        s[open + 1..close]
            .split([',', '/'])
            .filter(|p| !p.trim().is_empty())
            .filter_map(|p| p.trim().trim_end_matches('%').parse::<f64>().ok())
            .collect(),
    )
}

pub fn to_hsl(c: Rgb) -> (f64, f64, f64) {
    let (r, g, b) = (c.0 as f64 / 255.0, c.1 as f64 / 255.0, c.2 as f64 / 255.0);
    let max = r.max(g).max(b);
    let min = r.min(g).min(b);
    let l = (max + min) / 2.0;
    if max == min {
        return (0.0, 0.0, l); // grey: hue is undefined, report 0
    }
    let d = max - min;
    let s = if l > 0.5 { d / (2.0 - max - min) } else { d / (max + min) };
    let h = if max == r {
        (g - b) / d + if g < b { 6.0 } else { 0.0 }
    } else if max == g {
        (b - r) / d + 2.0
    } else {
        (r - g) / d + 4.0
    };
    (h * 60.0, s, l)
}

fn channel(p: f64, q: f64, t: f64) -> i32 {
    let t = (t % 1.0 + 1.0) % 1.0;
    let v = if t < 1.0 / 6.0 {
        p + (q - p) * 6.0 * t
    } else if t < 1.0 / 2.0 {
        q
    } else if t < 2.0 / 3.0 {
        p + (q - p) * (2.0 / 3.0 - t) * 6.0
    } else {
        p
    };
    round_half_even(v * 255.0) as i32
}

pub fn from_hsl(h: f64, s: f64, l: f64) -> Rgb {
    let h = (h % 360.0 + 360.0) % 360.0 / 360.0;
    let s = s.clamp(0.0, 1.0);
    let l = l.clamp(0.0, 1.0);
    let q = if l < 0.5 { l * (1.0 + s) } else { l + s - l * s };
    let p = 2.0 * l - q;
    (
        channel(p, q, h + 1.0 / 3.0),
        channel(p, q, h),
        channel(p, q, h - 1.0 / 3.0),
    )
}

pub fn hex(c: Rgb) -> String {
    format!("#{:02X}{:02X}{:02X}", c.0, c.1, c.2)
}

/// The same colour in every notation — what the tool is actually for.
pub fn describe_rgb(c: Rgb) -> String {
    let (h, s, l) = to_hsl(c);
    [
        format!("HEX  #{:02X}{:02X}{:02X}", c.0, c.1, c.2),
        format!("RGB  rgb({}, {}, {})", c.0, c.1, c.2),
        format!(
            "HSL  hsl({}, {}%, {}%)",
            round_half_away(h),
            round_half_away(s * 100.0),
            round_half_away(l * 100.0)
        ),
    ]
    .join("\n")
}

/// The C# side renders HSL with the "0" format, which rounds midpoints to even.
fn round_half_away(v: f64) -> i64 {
    round_half_even(v) as i64
}

pub fn describe(input: &str) -> Result<String, String> {
    Ok(describe_rgb(parse(input)?))
}

/// Colour harmonies built by rotating the hue.
pub fn palette(input: &str) -> Result<String, String> {
    let (h, s, l) = to_hsl(parse(input)?);
    let at = |degrees: f64| hex(from_hsl(h + degrees, s, l));
    Ok([
        format!("{:<16} {}", i18n::get("Color_Base"), at(0.0)),
        format!("{:<16} {}", i18n::get("Color_Complementary"), at(180.0)),
        format!("{:<16} {}  {}", i18n::get("Color_Triadic"), at(120.0), at(240.0)),
        format!("{:<16} {}  {}", i18n::get("Color_Analogous"), at(-30.0), at(30.0)),
        format!("{:<16} {}  {}", i18n::get("Color_SplitComp"), at(150.0), at(210.0)),
        format!(
            "{:<16} {}  {}  {}",
            i18n::get("Color_Tetradic"),
            at(90.0),
            at(180.0),
            at(270.0)
        ),
    ]
    .join("\n"))
}

fn relative_luminance(c: Rgb) -> f64 {
    // sRGB to linear, per the WCAG definition, then the Rec.709 luminance weights.
    fn ch(v: i32) -> f64 {
        let s = v as f64 / 255.0;
        if s <= 0.03928 { s / 12.92 } else { ((s + 0.055) / 1.055).powf(2.4) }
    }
    0.2126 * ch(c.0) + 0.7152 * ch(c.1) + 0.0722 * ch(c.2)
}

pub fn contrast_ratio(a: Rgb, b: Rgb) -> f64 {
    let (la, lb) = (relative_luminance(a), relative_luminance(b));
    let (hi, lo) = (la.max(lb), la.min(lb));
    (hi + 0.05) / (lo + 0.05)
}

/// WCAG contrast ratio between two colours (one per line) and which levels it passes.
pub fn contrast(input: &str) -> Result<String, String> {
    let lines: Vec<&str> = input
        .split('\n')
        .map(str::trim)
        .filter(|l| !l.is_empty())
        .collect();
    if lines.len() < 2 {
        return Err(i18n::get("Error_NeedTwoColors").to_string());
    }
    let ratio = contrast_ratio(parse(lines[0])?, parse(lines[1])?);
    let verdict = |min: f64| {
        if ratio >= min { i18n::get("Color_Pass") } else { i18n::get("Color_Fail") }
    };
    let (normal, large) = (i18n::get("Color_NormalText"), i18n::get("Color_LargeText"));
    Ok([
        format!("{}  {ratio:.2}:1", i18n::get("Color_Ratio")),
        format!("AA  ({normal})   {}", verdict(4.5)),
        format!("AA  ({large})   {}", verdict(3.0)),
        format!("AAA ({normal})   {}", verdict(7.0)),
        format!("AAA ({large})   {}", verdict(4.5)),
    ]
    .join("\n"))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_every_notation_to_the_same_colour() {
        assert_eq!(parse("#FF0000").unwrap(), (255, 0, 0));
        assert_eq!(parse("#f00").unwrap(), (255, 0, 0), "shorthand expands");
        assert_eq!(parse("ff0000").unwrap(), (255, 0, 0), "hash is optional");
        assert_eq!(parse("rgb(255, 0, 0)").unwrap(), (255, 0, 0));
        assert_eq!(parse("hsl(0, 100%, 50%)").unwrap(), (255, 0, 0));
    }

    #[test]
    fn rejects_nonsense() {
        let _guard = crate::i18n::test_lock();
        i18n::set_language("en");
        for bad in ["", "#12345", "zzz", "#gggggg"] {
            assert!(parse(bad).is_err(), "{bad} should not parse");
        }
    }

    #[test]
    fn describes_a_colour_in_every_notation() {
        let red = describe_rgb((255, 0, 0));
        assert!(red.contains("HEX  #FF0000"), "{red}");
        assert!(red.contains("RGB  rgb(255, 0, 0)"), "{red}");
        assert!(red.contains("HSL  hsl(0, 100%, 50%)"), "{red}");
        assert!(describe_rgb((128, 128, 128)).contains("HSL  hsl(0, 0%, 50%)"), "grey has no hue");
    }

    #[test]
    fn hsl_round_trips_through_rgb() {
        for c in [(255, 0, 0), (0, 255, 0), (0, 0, 255), (18, 52, 86), (255, 255, 255)] {
            let (h, s, l) = to_hsl(c);
            assert_eq!(from_hsl(h, s, l), c, "round trip failed for {c:?}");
        }
    }

    /// Black on white is the reference maximum in WCAG.
    #[test]
    fn contrast_ratio_matches_the_wcag_reference() {
        let ratio = contrast_ratio((0, 0, 0), (255, 255, 255));
        assert!((ratio - 21.0).abs() < 0.01, "black on white should be 21:1, got {ratio}");
        assert!((contrast_ratio((255, 0, 0), (255, 0, 0)) - 1.0).abs() < 1e-9, "identical is 1:1");
    }

    #[test]
    fn contrast_needs_two_colours() {
        let _guard = crate::i18n::test_lock();
        i18n::set_language("en");
        assert!(contrast("#000000").is_err());
        assert!(contrast("#000000\n#FFFFFF").is_ok());
    }

    #[test]
    fn palette_rotates_the_hue() {
        let _guard = crate::i18n::test_lock();
        i18n::set_language("en");
        let result = palette("#FF0000").unwrap();
        assert_eq!(result.lines().count(), 6);
        assert!(result.contains("#FF0000"), "base colour is present: {result}");
        // The complement of red is cyan.
        assert!(result.contains("#00FFFF"), "{result}");
    }
}
